File: kubarr/api/routers/logs.py
"""Logs API router."""

import httpx
from datetime import datetime, timedelta
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel

from kubarr.api.dependencies import get_current_active_user, get_k8s_client, get_logs_service
from kubarr.core.k8s_client import K8sClientManager
from kubarr.core.logs_service import LogsService
from kubarr.core.models import LogEntry, LogFilter
from kubarr.core.models_auth import User

# Authentication via oauth2-proxy headers (X-Auth-Request-User)
router = APIRouter(dependencies=[Depends(get_current_active_user)])

# Loki service URL (inside cluster)
LOKI_URL = "http://loki.loki.svc.cluster.local:3100"


class LokiLogEntry(BaseModel):
    """A log entry from Loki."""
    timestamp: str
    line: str
    labels: dict


class LokiQueryResponse(BaseModel):
    """Response from Loki query."""
    streams: List[dict]
    total_entries: int


@router.get("/{pod_name}", response_model=List[LogEntry])
async def get_pod_logs(
    pod_name: str,
    namespace: str = Query(default="media", description="Namespace"),
    container: Optional[str] = Query(default=None, description="Container name"),
    tail: int = Query(default=100, description="Number of lines"),
    k8s_client: K8sClientManager = Depends(get_k8s_client)
) -> List[LogEntry]:
    """Get logs from a specific pod.

    Args:
        pod_name: Pod name
        namespace: Namespace
        container: Optional container name
        tail: Number of lines to return

    Returns:
        List of log entries

    Raises:
        HTTPException: If log retrieval fails
    """
    try:
        service = LogsService(k8s_client=k8s_client)
        log_filter = LogFilter(
            namespace=namespace,
            pod_name=pod_name,
            container=container,
            tail_lines=tail
        )
        return service.get_logs(log_filter)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/app/{app_name}", response_model=List[LogEntry])
async def get_app_logs(
    app_name: str,
    namespace: str = Query(default="media", description="Namespace"),
    tail: int = Query(default=100, description="Number of lines per pod"),
    k8s_client: K8sClientManager = Depends(get_k8s_client)
) -> List[LogEntry]:
    """Get logs from all pods of an app.

    Args:
        app_name: App name
        namespace: Namespace
        tail: Number of lines to return per pod

    Returns:
        List of log entries from all pods

    Raises:
        HTTPException: If log retrieval fails
    """
    try:
        service = LogsService(k8s_client=k8s_client)
        log_filter = LogFilter(
            namespace=namespace,
            app_name=app_name,
            tail_lines=tail
        )
        return service.get_logs(log_filter)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/raw/{pod_name}", response_model=str)
async def get_raw_pod_logs(
    pod_name: str,
    namespace: str = Query(default="media", description="Namespace"),
    container: Optional[str] = Query(default=None, description="Container name"),
    tail: int = Query(default=100, description="Number of lines"),
    k8s_client: K8sClientManager = Depends(get_k8s_client)
) -> str:
    """Get raw logs from a pod as plain text.

    Args:
        pod_name: Pod name
        namespace: Namespace
        container: Optional container name
        tail: Number of lines

    Returns:
        Raw log string

    Raises:
        HTTPException: If log retrieval fails
    """
    try:
        service = LogsService(k8s_client=k8s_client)
        return service.get_pod_logs(
            pod_name=pod_name,
            namespace=namespace,
            container=container,
            tail=tail
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


# ============== Loki Endpoints ==============

@router.get("/loki/labels")
async def get_loki_labels():
    """Get all available labels from Loki.

    Returns:
        List of label names
    """
    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.get(f"{LOKI_URL}/loki/api/v1/labels")
            response.raise_for_status()
            data = response.json()
            return data.get("data", [])
    except httpx.RequestError as e:
        raise HTTPException(status_code=503, detail=f"Failed to connect to Loki: {str(e)}")
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/loki/label/{label}/values")
async def get_loki_label_values(label: str):
    """Get all values for a specific label from Loki.

    Args:
        label: Label name (e.g., 'namespace', 'app', 'container')

    Returns:
        List of label values
    """
    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.get(f"{LOKI_URL}/loki/api/v1/label/{label}/values")
            response.raise_for_status()
            data = response.json()
            return data.get("data", [])
    except httpx.RequestError as e:
        raise HTTPException(status_code=503, detail=f"Failed to connect to Loki: {str(e)}")
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/loki/query")
async def query_loki_logs(
    query: str = Query(default='{namespace=~".+"}', description="LogQL query"),
    start: Optional[str] = Query(default=None, description="Start time (RFC3339 or Unix nanoseconds)"),
    end: Optional[str] = Query(default=None, description="End time (RFC3339 or Unix nanoseconds)"),
    limit: int = Query(default=1000, description="Maximum number of entries to return"),
    direction: str = Query(default="backward", description="Log order: 'forward' or 'backward'"),
):
    """Query logs from Loki using LogQL.

    Args:
        query: LogQL query string (e.g., '{namespace="sonarr"}')
        start: Start time for the query range
        end: End time for the query range
        limit: Maximum entries to return
        direction: Log order (backward = newest first)

    Returns:
        Log streams with entries
    """
    try:
        # Default to last hour if no time range specified
        now = datetime.utcnow()
        if not end:
            end = str(int(now.timestamp() * 1e9))
        if not start:
            start_time = now - timedelta(hours=1)
            start = str(int(start_time.timestamp() * 1e9))

        params = {
            "query": query,
            "start": start,
            "end": end,
            "limit": limit,
            "direction": direction,
        }

        async with httpx.AsyncClient(timeout=60.0) as client:
            response = await client.get(
                f"{LOKI_URL}/loki/api/v1/query_range",
                params=params
            )
            response.raise_for_status()
            data = response.json()

            # Parse Loki response
            result = data.get("data", {}).get("result", [])
            streams = []
            total_entries = 0

            for stream in result:
                labels = stream.get("stream", {})
                values = stream.get("values", [])
                entries = []

                for value in values:
                    timestamp_ns, line = value
                    # Convert nanoseconds to ISO format
                    timestamp_s = int(timestamp_ns) / 1e9
                    dt = datetime.utcfromtimestamp(timestamp_s)
                    entries.append({
                        "timestamp": dt.isoformat() + "Z",
                        "line": line,
                    })
                    total_entries += 1

                streams.append({
                    "labels": labels,
                    "entries": entries,
                })

            return {
                "streams": streams,
                "total_entries": total_entries,
            }

    except httpx.RequestError as e:
        raise HTTPException(status_code=503, detail=f"Failed to connect to Loki: {str(e)}")
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/loki/namespaces")
async def get_loki_namespaces():
    """Get all namespaces that have logs in Loki.

    Returns:
        List of namespace names
    """
    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.get(f"{LOKI_URL}/loki/api/v1/label/namespace/values")
            response.raise_for_status()
            data = response.json()
            return data.get("data", [])
    except httpx.RequestError as e:
        raise HTTPException(status_code=503, detail=f"Failed to connect to Loki: {str(e)}")
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: code/backend/src/endpoints/logs.rs
use axum::{
    extract::{Path, Query, State},
    routing::get,
    Json, Router,
};
use chrono::{DateTime, Duration, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

use crate::middleware::permissions::{Authorized, LogsView};
use crate::error::{AppError, Result};
use crate::state::AppState;

// Loki service URL inside the cluster
const LOKI_URL: &str = "http://loki.loki.svc.cluster.local:3100";

pub fn logs_routes(state: AppState) -> Router {
    Router::new()
        // Loki endpoints (must be before /:pod_name to avoid conflicts)
        .route("/loki/namespaces", get(get_loki_namespaces))
        .route("/loki/labels", get(get_loki_labels))
        .route("/loki/label/:label/values", get(get_loki_label_values))
        .route("/loki/query", get(query_loki_logs))
        // Pod logs endpoints
        .route("/raw/:pod_name", get(get_raw_pod_logs))
        .route("/app/:app_name", get(get_app_logs))
        .route("/:pod_name", get(get_pod_logs))
        .with_state(state)
}

#[derive(Debug, Serialize)]
struct LogEntry {
    timestamp: String,
    line: String,
    pod_name: String,
    container: Option<String>,
}

#[derive(Debug, Deserialize)]
struct PodLogsQuery {
    #[serde(default = "default_namespace")]
    namespace: String,
    container: Option<String>,
    #[serde(default = "default_tail")]
    tail: i32,
}

fn default_namespace() -> String {
    "media".to_string()
}

fn default_tail() -> i32 {
    100
}

#[derive(Debug, Deserialize)]
struct LokiQueryParams {
    #[serde(default = "default_loki_query")]
    query: String,
    start: Option<String>,
    end: Option<String>,
    #[serde(default = "default_limit")]
    limit: i32,
    #[serde(default = "default_direction")]
    direction: String,
}

fn default_loki_query() -> String {
    "{namespace=~\".+\"}".to_string()
}

fn default_limit() -> i32 {
    1000
}

fn default_direction() -> String {
    "backward".to_string()
}

#[derive(Debug, Serialize)]
struct LokiQueryResponse {
    streams: Vec<LokiStream>,
    total_entries: i32,
}

#[derive(Debug, Serialize)]
struct LokiStream {
    labels: HashMap<String, String>,
    entries: Vec<LokiEntry>,
}

#[derive(Debug, Serialize)]
struct LokiEntry {
    timestamp: String,
    line: String,
}

/// Get logs from a specific pod
async fn get_pod_logs(
    State(state): State<AppState>,
    Path(pod_name): Path<String>,
    Query(params): Query<PodLogsQuery>,
    _auth: Authorized<LogsView>,
) -> Result<Json<Vec<LogEntry>>> {
    let k8s = state.k8s_client.read().await;
    let client = k8s
        .as_ref()
        .ok_or_else(|| AppError::Internal("Kubernetes client not available".to_string()))?;

    let logs = client
        .get_pod_logs(
            &pod_name,
            &params.namespace,
            params.container.as_deref(),
            params.tail,
        )
        .await?;

    let entries: Vec<LogEntry> = logs
        .lines()
        .map(|line| LogEntry {
            timestamp: Utc::now().to_rfc3339(),
            line: line.to_string(),
            pod_name: pod_name.clone(),
            container: params.container.clone(),
        })
        .collect();

    Ok(Json(entries))
}

/// Get logs from all pods of an app
async fn get_app_logs(
    State(state): State<AppState>,
    Path(app_name): Path<String>,
    Query(params): Query<PodLogsQuery>,
    _auth: Authorized<LogsView>,
) -> Result<Json<Vec<LogEntry>>> {
    let k8s = state.k8s_client.read().await;
    let client = k8s
        .as_ref()
        .ok_or_else(|| AppError::Internal("Kubernetes client not available".to_string()))?;

    // Get all pods for the app
    let pods = client.list_pods(&params.namespace, Some(&app_name)).await?;

    let mut all_entries = Vec::new();

    for pod in pods {
        if let Some(pod_name) = pod.metadata.name {
            match client
                .get_pod_logs(
                    &pod_name,
                    &params.namespace,
                    params.container.as_deref(),
                    params.tail,
                )
                .await
            {
                Ok(logs) => {
                    for line in logs.lines() {
                        all_entries.push(LogEntry {
                            timestamp: Utc::now().to_rfc3339(),
                            line: line.to_string(),
                            pod_name: pod_name.clone(),
                            container: params.container.clone(),
                        });
                    }
                }
                Err(e) => {
                    tracing::warn!("Failed to get logs for pod {}: {}", pod_name, e);
                }
            }
        }
    }

    Ok(Json(all_entries))
}

/// Get raw logs from a pod as plain text
async fn get_raw_pod_logs(
    State(state): State<AppState>,
    Path(pod_name): Path<String>,
    Query(params): Query<PodLogsQuery>,
    _auth: Authorized<LogsView>,
) -> Result<String> {
    let k8s = state.k8s_client.read().await;
    let client = k8s
        .as_ref()
        .ok_or_else(|| AppError::Internal("Kubernetes client not available".to_string()))?;

    let logs = client
        .get_pod_logs(
            &pod_name,
            &params.namespace,
            params.container.as_deref(),
            params.tail,
        )
        .await?;

    Ok(logs)
}

// ============== Loki Endpoints ==============

/// Get all namespaces that have logs in Loki
async fn get_loki_namespaces(
    _auth: Authorized<LogsView>,
) -> Result<Json<Vec<String>>> {
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(30))
        .build()
        .map_err(|e| AppError::Internal(format!("Failed to create HTTP client: {}", e)))?;

    let response = client
        .get(format!("{}/loki/api/v1/label/namespace/values", LOKI_URL))
        .send()
        .await
        .map_err(|e| AppError::ServiceUnavailable(format!("Failed to connect to Loki: {}", e)))?;

    if !response.status().is_success() {
        return Err(AppError::Internal(format!(
            "Loki returned error: {}",
            response.status()
        )));
    }

    let data: serde_json::Value = response
        .json()
        .await
        .map_err(|e| AppError::Internal(format!("Failed to parse Loki response: {}", e)))?;

    let namespaces = data
        .get("data")
        .and_then(|d| d.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|v| v.as_str().map(String::from))
                .collect()
        })
        .unwrap_or_default();

    Ok(Json(namespaces))
}

/// Get all available labels from Loki
async fn get_loki_labels(
    _auth: Authorized<LogsView>,
) -> Result<Json<Vec<String>>> {
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(30))
        .build()
        .map_err(|e| AppError::Internal(format!("Failed to create HTTP client: {}", e)))?;

    let response = client
        .get(format!("{}/loki/api/v1/labels", LOKI_URL))
        .send()
        .await
        .map_err(|e| AppError::ServiceUnavailable(format!("Failed to connect to Loki: {}", e)))?;

    if !response.status().is_success() {
        return Err(AppError::Internal(format!(
            "Loki returned error: {}",
            response.status()
        )));
    }

    let data: serde_json::Value = response
        .json()
        .await
        .map_err(|e| AppError::Internal(format!("Failed to parse Loki response: {}", e)))?;

    let labels = data
        .get("data")
        .and_then(|d| d.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|v| v.as_str().map(String::from))
                .collect()
        })
        .unwrap_or_default();

    Ok(Json(labels))
}

/// Get all values for a specific label from Loki
async fn get_loki_label_values(
    Path(label): Path<String>,
    _auth: Authorized<LogsView>,
) -> Result<Json<Vec<String>>> {
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(30))
        .build()
        .map_err(|e| AppError::Internal(format!("Failed to create HTTP client: {}", e)))?;

    let response = client
        .get(format!("{}/loki/api/v1/label/{}/values", LOKI_URL, label))
        .send()
        .await
        .map_err(|e| AppError::ServiceUnavailable(format!("Failed to connect to Loki: {}", e)))?;

    if !response.status().is_success() {
        return Err(AppError::Internal(format!(
            "Loki returned error: {}",
            response.status()
        )));
    }

    let data: serde_json::Value = response
        .json()
        .await
        .map_err(|e| AppError::Internal(format!("Failed to parse Loki response: {}", e)))?;

    let values = data
        .get("data")
        .and_then(|d| d.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|v| v.as_str().map(String::from))
                .collect()
        })
        .unwrap_or_default();

    Ok(Json(values))
}

/// Query logs from Loki using LogQL
async fn query_loki_logs(
    Query(params): Query<LokiQueryParams>,
    _auth: Authorized<LogsView>,
) -> Result<Json<LokiQueryResponse>> {
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(60))
        .build()
        .map_err(|e| AppError::Internal(format!("Failed to create HTTP client: {}", e)))?;

    // Default to last hour if no time range specified
    let now = Utc::now();
    let end = params
        .end
        .unwrap_or_else(|| format!("{}", now.timestamp_nanos_opt().unwrap_or(0)));
    let start = params.start.unwrap_or_else(|| {
        let start_time = now - Duration::hours(1);
        format!("{}", start_time.timestamp_nanos_opt().unwrap_or(0))
    });

    let response = client
        .get(format!("{}/loki/api/v1/query_range", LOKI_URL))
        .query(&[
            ("query", params.query.as_str()),
            ("start", start.as_str()),
            ("end", end.as_str()),
            ("limit", &params.limit.to_string()),
            ("direction", params.direction.as_str()),
        ])
        .send()
        .await
        .map_err(|e| AppError::ServiceUnavailable(format!("Failed to connect to Loki: {}", e)))?;

    if !response.status().is_success() {
        return Err(AppError::Internal(format!(
            "Loki returned error: {}",
            response.status()
        )));
    }

    let data: serde_json::Value = response
        .json()
        .await
        .map_err(|e| AppError::Internal(format!("Failed to parse Loki response: {}", e)))?;

    // Parse Loki response
    let result = data
        .get("data")
        .and_then(|d| d.get("result"))
        .and_then(|r| r.as_array())
        .cloned()
        .unwrap_or_default();

    let mut streams = Vec::new();
    let mut total_entries = 0;

    for stream in result {
        let labels: HashMap<String, String> = stream
            .get("stream")
            .and_then(|s| serde_json::from_value(s.clone()).ok())
            .unwrap_or_default();

        let values = stream
            .get("values")
            .and_then(|v| v.as_array())
            .cloned()
            .unwrap_or_default();

        let mut entries = Vec::new();

        for value in values {
            if let Some(arr) = value.as_array() {
                if arr.len() >= 2 {
                    let timestamp_ns = arr[0].as_str().unwrap_or("0");
                    let line = arr[1].as_str().unwrap_or("").to_string();

                    // Convert nanoseconds to ISO format
                    if let Ok(ns) = timestamp_ns.parse::<i64>() {
                        let secs = ns / 1_000_000_000;
                        let nsecs = (ns % 1_000_000_000) as u32;
                        if let Some(dt) = DateTime::from_timestamp(secs, nsecs) {
                            entries.push(LokiEntry {
                                timestamp: dt.to_rfc3339(),
                                line,
                            });
                            total_entries += 1;
                        }
                    }
                }
            }
        }

        streams.push(LokiStream { labels, entries });
    }

    Ok(Json(LokiQueryResponse {
        streams,
        total_entries,
    }))
}
